File: apps/api/internal/shooter/simulation.go
package shooter

import (
	"fmt"
	"slices"
)

const (
	playerY      = 5200
	playerRadius = 95
	enemyRadius  = 120
	bulletRadius = 42
	grazeRadius  = 180
)

type enemyEntity struct {
	id, specIndex, x, y, health, maxHealth int
	fireClock, age, phase, warning, volley int
	marks                                  int
	boss                                   bool
}

type projectileEntity struct {
	id, x, y, vx, vy, damage, pierce int
	radius, width, health            int
	kind                             string
	hostile, grazed                  bool
}

type pickupEntity struct {
	id, x, y, value int
}

type effectEntity struct {
	id, x, y, ticks, power int
	kind                   string
}

type runtime struct {
	damage, fireInterval, multishot, pierce    int
	startingShield, maxHealth                  int
	rescueCharge, rescueDamage                 int
	companionPower, grazeCharge                int
	spread, guardOnSpecial, pickupMagnet       int
	echoVolley, bossBreak, lowHealthPower      int
	comboExtend, companionCharge, recoveryDrop int
}

type simulation struct {
	config             Config
	random             randomStream
	runtime            runtime
	tick               int
	playerX            int
	health             int
	shield             int
	invulnerableTicks  int
	rescueCharge       int
	rescueHeld         bool
	rescuesUsed        int
	grazeCount         int
	combo              int
	comboClock         int
	kills              int
	score              int
	attackClock        int
	attackSequence     int
	alignmentTicks     int
	companionClocks    []int
	nextEnemyID        int
	nextProjectileID   int
	nextPickupID       int
	nextEffectID       int
	spawnedBoss        bool
	lastRescueTick     int
	bossPhaseTick      int
	dailyVariant       string
	enemies            []enemyEntity
	enemyProjectiles   []projectileEntity
	playerProjectiles  []projectileEntity
	pickups            []pickupEntity
	pickupsCollected   int
	lastPickupTick     int
	pressureQuietTicks int
	effects            []effectEntity
}

func Simulate(config Config, trace InputTrace) (Result, error) {
	if err := normalizeConfig(&config); err != nil {
		return Result{}, err
	}
	frames, err := DecodeTrace(trace, config.DurationTicks)
	if err != nil {
		return Result{}, err
	}
	sim := newSimulation(config)
	for _, frame := range frames {
		sim.step(frame)
	}
	won := sim.health > 0
	if config.Boss != nil {
		won = won && !slices.ContainsFunc(sim.enemies, func(enemy enemyEntity) bool { return enemy.boss && enemy.health > 0 })
	}
	return sim.result(won), nil
}

func normalizeConfig(config *Config) error {
	if config.Seed == "" || config.DurationTicks <= 0 || config.DurationTicks > MaxSegmentTicks || config.EncoreLevel < 0 || config.EncoreLevel > 3 {
		return fmt.Errorf("shooter: invalid segment configuration")
	}
	if config.StoryChoiceID != "" && storyChoiceMode(config.StoryChoiceID) == 0 {
		return fmt.Errorf("shooter: invalid story choice")
	}
	if !slices.Contains(SupportedKits[:], config.Kit.ID) || config.Kit.MaxHealth != 3 || config.Kit.AttackDamage <= 0 || config.Kit.FireInterval <= 0 || config.Kit.RescueDamage <= 0 {
		return fmt.Errorf("shooter: invalid character kit")
	}
	if config.PlayerHealth <= 0 || config.PlayerHealth > config.Kit.MaxHealth || config.StartingRescueCharge < 0 || config.StartingRescueCharge > RescueChargeLimit {
		return fmt.Errorf("shooter: invalid player health")
	}
	if config.SpecialChargePenaltyPercent < 0 || config.SpecialChargePenaltyPercent > 75 {
		return fmt.Errorf("shooter: invalid special charge penalty")
	}
	if config.Limits == (Limits{}) {
		config.Limits = DefaultLimits()
	}
	if config.Limits.Enemies <= 0 || config.Limits.Enemies > MaxEnemies || config.Limits.EnemyProjectiles <= 0 || config.Limits.EnemyProjectiles > MaxEnemyBullets || config.Limits.PlayerProjectiles <= 0 || config.Limits.PlayerProjectiles > MaxPlayerBullets || config.Limits.Pickups <= 0 || config.Limits.Pickups > MaxPickups || config.Limits.Effects <= 0 || config.Limits.Effects > MaxEffects {
		return fmt.Errorf("shooter: invalid entity limits")
	}
	seenEnemies := make(map[string]bool, len(config.Enemies))
	for _, enemy := range config.Enemies {
		if enemy.ID == "" || seenEnemies[enemy.ID] || !slices.Contains(SupportedChassis[:], enemy.Chassis) || enemy.Health <= 0 || enemy.Speed < 0 || enemy.FireInterval <= 0 || enemy.ProjectileSpeed <= 0 || enemy.Damage <= 0 {
			return fmt.Errorf("shooter: invalid enemy %q", enemy.ID)
		}
		seenEnemies[enemy.ID] = true
	}
	validFormations := []string{"line", "fan", "staggered", "pincer", "center", "sweep"}
	for _, spawn := range config.Wave.Spawns {
		if spawn.AtTick < 0 || spawn.AtTick >= config.DurationTicks || !seenEnemies[spawn.EnemyID] || spawn.Count <= 0 || spawn.IntervalTicks < 0 || !slices.Contains(validFormations, spawn.Formation) {
			return fmt.Errorf("shooter: invalid wave spawn")
		}
	}
	seenCompanions := make(map[CompanionID]bool, len(config.Companions))
	validTriggers := []string{"segment_start", "graze_streak", "low_health", "special_used", "boss_stage", "pickup_chain", "wave_clear"}
	validAssists := []string{"side_shot", "shield", "echo_shot", "clear_lane", "convert_bullet", "focus_beam", "heal"}
	for _, companion := range config.Companions {
		if !slices.Contains(SupportedCompanions[:], companion.ID) || seenCompanions[companion.ID] || !slices.Contains(validTriggers, companion.Trigger) || !slices.Contains(validAssists, companion.Behavior) || companion.Amount <= 0 || companion.CooldownTicks < 0 {
			return fmt.Errorf("shooter: invalid companion")
		}
		seenCompanions[companion.ID] = true
	}
	for _, effect := range config.ShowEffects {
		if !slices.Contains(SupportedShowEffects[:], effect.Kind) || effect.Amount <= 0 {
			return fmt.Errorf("shooter: invalid show effect %q", effect.Kind)
		}
	}
	if config.Boss != nil {
		if !slices.Contains(SupportedBosses[:], config.Boss.ID) || config.Boss.Health <= 0 || len(config.Boss.Stages) != 3 {
			return fmt.Errorf("shooter: invalid boss")
		}
		lastThreshold := 101
		for _, stage := range config.Boss.Stages {
			if stage.HealthThreshold < 0 || stage.HealthThreshold >= lastThreshold || stage.FireInterval <= 0 || stage.ProjectileSpeed <= 0 || stage.Damage <= 0 || stage.TelegraphTicks < 0 || stage.TelegraphTicks >= stage.FireInterval {
				return fmt.Errorf("shooter: invalid boss stage")
			}
			lastThreshold = stage.HealthThreshold
		}
	}
	return nil
}

func newSimulation(config Config) *simulation {
	resolved := resolveRuntime(config)
	variant := ""
	if config.Daily {
		variant = config.DailyModifierID
	}
	sim := &simulation{
		config: config, random: randomStream{state: seedFromString(config.Seed)}, runtime: resolved,
		playerX: ArenaWidth / 2, health: config.PlayerHealth, shield: resolved.startingShield,
		rescueCharge: clamp(resolved.rescueCharge, 0, RescueChargeLimit), dailyVariant: variant,
		companionClocks:   make([]int, len(config.Companions)),
		enemies:           make([]enemyEntity, 0, config.Limits.Enemies),
		enemyProjectiles:  make([]projectileEntity, 0, config.Limits.EnemyProjectiles),
		playerProjectiles: make([]projectileEntity, 0, config.Limits.PlayerProjectiles),
		pickups:           make([]pickupEntity, 0, config.Limits.Pickups),
		effects:           make([]effectEntity, 0, config.Limits.Effects),
	}
	for index, companion := range config.Companions {
		if companion.Trigger == "segment_start" {
			sim.companionClocks[index] = max(1, companion.CooldownTicks)
		}
	}
	return sim
}

func resolveRuntime(config Config) runtime {
	result := runtime{
		damage: config.Kit.AttackDamage, fireInterval: config.Kit.FireInterval,
		multishot: 1, maxHealth: config.Kit.MaxHealth, startingShield: config.Kit.StartingShield,
		rescueCharge: config.StartingRescueCharge, rescueDamage: config.Kit.RescueDamage, grazeCharge: 4,
	}
	for _, effect := range config.ShowEffects {
		switch effect.Kind {
		case EffectTwinShot:
			result.multishot += effect.Amount
		case EffectPiercingShot:
			result.pierce += effect.Amount
		case EffectSpreadShot:
			result.spread += effect.Amount
			result.multishot += 2
		case EffectGrazeCharge:
			result.grazeCharge += effect.Amount
		case EffectGuardOnSpecial:
			result.guardOnSpecial += effect.Amount
		case EffectPickupMagnet:
			result.pickupMagnet += effect.Amount
		case EffectEchoVolley:
			result.echoVolley += effect.Amount
		case EffectBossBreak:
			result.bossBreak += effect.Amount
		case EffectLowHealthPower:
			result.lowHealthPower += effect.Amount
		case EffectComboExtend:
			result.comboExtend += effect.Amount
		case EffectCompanionCharge:
			result.companionCharge += effect.Amount
		case EffectRecoveryDrop:
			result.recoveryDrop += effect.Amount
		}
	}
	result.fireInterval = max(3, result.fireInterval)
	result.multishot = clamp(result.multishot, 1, 5)
	return result
}

func (sim *simulation) step(input Input) {
	sim.tick++
	if sim.invulnerableTicks > 0 {
		sim.invulnerableTicks--
	}
	moveLimit := sim.config.Kit.MoveLimit
	if moveLimit <= 0 || moveLimit > ArenaWidth/2-playerRadius {
		moveLimit = ArenaWidth/2 - playerRadius
	}
	sim.playerX = ArenaWidth/2 - moveLimit + int(input.X)*(moveLimit*2)/127
	if input.Rescue && !sim.rescueHeld {
		sim.activateRescue()
	}
	sim.rescueHeld = input.Rescue
	sim.spawnWave()
	sim.spawnBoss()
	sim.updateWeapons()
	sim.updateCompanions()
	sim.updateEnemies()
	sim.updateKitPassives()
	sim.updateProjectiles()
	sim.updatePickups()
	sim.updateEffects()
	if sim.comboClock > 0 {
		sim.comboClock--
	} else {
		sim.combo = 0
	}
	if sim.health < 0 {
		sim.health = 0
	}
}

func (sim *simulation) result(won bool) Result {
	if won {
		sim.score += sim.health*10 + sim.rescueCharge*2
	}
	final := sim.snapshot()
	final.Score = sim.score
	return Result{Won: won, Health: sim.health, Ticks: sim.tick, Kills: sim.kills, RescuesUsed: sim.rescuesUsed, Grazes: sim.grazeCount, Score: sim.score, DailyVariant: sim.dailyVariant, Final: final}
}

func (sim *simulation) snapshot() Snapshot {
	enemies := make([]EnemySnapshot, 0, len(sim.enemies))
	for _, enemy := range sim.enemies {
		var spec EnemySpec
		id, chassis, intent := "", ChassisSpamBot, ""
		if enemy.boss && sim.config.Boss != nil {
			id, chassis = string(sim.config.Boss.ID), ChassisCensorFrame
			if stage := enemy.phase - 1; stage >= 0 && stage < len(sim.config.Boss.Stages) && sim.config.Boss.Stages[stage].TelegraphTicks > 0 && enemy.fireClock >= max(1, sim.config.Boss.Stages[stage].FireInterval-sim.config.Boss.Stages[stage].TelegraphTicks) {
				intent = "fire"
			}
		} else {
			spec = sim.config.Enemies[enemy.specIndex]
			id, chassis, intent = spec.ID, spec.Chassis, enemyIntent(enemy, spec, sim.config.EncoreLevel)
		}
		enemies = append(enemies, EnemySnapshot{ID: enemy.id, SpecID: id, Chassis: chassis, Position: Position{X: enemy.x, Y: enemy.y}, Health: max(0, enemy.health), MaxHealth: enemy.maxHealth, Boss: enemy.boss, Stage: enemy.phase, Intent: intent, Marks: enemy.marks})
	}
	return Snapshot{
		Tick: sim.tick, PlayerX: sim.playerX, Health: sim.health, MaxHealth: sim.runtime.maxHealth,
		Shield: sim.shield, InvulnerableTicks: sim.invulnerableTicks, RescueCharge: sim.rescueCharge, RescuesUsed: sim.rescuesUsed,
		GrazeCount: sim.grazeCount, Combo: sim.combo, Score: sim.score, DailyVariant: sim.dailyVariant,
		Enemies: enemies, EnemyProjectiles: projectileSnapshots(sim.enemyProjectiles), PlayerProjectiles: projectileSnapshots(sim.playerProjectiles),
		Pickups: pickupSnapshots(sim.pickups), Threats: sim.threatSnapshots(), Effects: effectSnapshots(sim.effects),
	}
}

func projectileSnapshots(entities []projectileEntity) []ProjectileSnapshot {
	result := make([]ProjectileSnapshot, 0, len(entities))
	for _, item := range entities {
		result = append(result, ProjectileSnapshot{
			ID: item.id, Position: Position{X: item.x, Y: item.y}, Velocity: Position{X: item.vx, Y: item.vy}, Hostile: item.hostile,
			Kind: item.kind, Radius: item.radius, Width: item.width, Health: max(0, item.health),
		})
	}
	return result
}

func pickupSnapshots(entities []pickupEntity) []PickupSnapshot {
	result := make([]PickupSnapshot, 0, len(entities))
	for _, item := range entities {
		result = append(result, PickupSnapshot{ID: item.id, Kind: "support_note", Position: Position{X: item.x, Y: item.y}, Value: item.value})
	}
	return result
}

func effectSnapshots(entities []effectEntity) []EffectSnapshot {
	result := make([]EffectSnapshot, 0, len(entities))
	for _, item := range entities {
		result = append(result, EffectSnapshot{ID: item.id, Kind: item.kind, Position: Position{X: item.x, Y: item.y}, Ticks: item.ticks, Power: item.power})
	}
	return result
}

func (sim *simulation) addEffect(kind string, x, y, ticks, power int) {
	if kind == "" || ticks <= 0 || len(sim.effects) >= sim.config.Limits.Effects {
		return
	}
	sim.nextEffectID++
	sim.effects = append(sim.effects, effectEntity{id: sim.nextEffectID, kind: kind, x: x, y: y, ticks: ticks, power: power})
}

func (sim *simulation) updateEffects() {
	kept := sim.effects[:0]
	for _, effect := range sim.effects {
		effect.ticks--
		if effect.ticks > 0 {
			kept = append(kept, effect)
		}
	}
	sim.effects = kept
}

File: apps/miniapp/src/features/shooter/simulation.ts
import {
  ENEMY_RADIUS,
  PLAYER_RADIUS,
  SHOOTER_WIDTH,
  clamp,
} from "@/features/shooter/constants";
import { spawnBoss, updateBoss } from "@/features/shooter/bosses";
import {
  damagePlayer,
  encoreInterval,
  enemyIntent,
  fireEnemy,
  fireEnemySecondary,
  hasTrait,
  moveEnemy,
  removeDefeatedEnemies,
  threatSnapshots,
  updateEffects,
  updateKitPassives,
  updatePickups,
  updateProjectiles,
} from "@/features/shooter/enemies";
import type { ShooterInput } from "@/features/shooter/input";
import { ShooterRandom } from "@/features/shooter/random";
import { activateRescue } from "@/features/shooter/specials";
import type {
  ShooterMutableState,
  ShooterResult,
  ShooterRuntime,
  ShooterSnapshot,
  ShooterStepEvents,
} from "@/features/shooter/types";
import { emptyStepEvents } from "@/features/shooter/types";
import { createShooterRuntime, updateCompanions, updateWeapons } from "@/features/shooter/weapons";
import { spawnWave } from "@/features/shooter/waves";
import type { ShooterRuntimeConfig } from "@/lib/api/types";

export { createShooterRuntime } from "@/features/shooter/weapons";

export type ShooterSimulation = {
  readonly snapshot: () => ShooterSnapshot;
  readonly step: (input: ShooterInput) => ShooterStepEvents;
  readonly result: () => ShooterResult | null;
};

const createInitialState = (runtime: ShooterRuntime): ShooterMutableState => ({
  config: runtime.config,
  runtime: runtime.resolved,
  random: new ShooterRandom(runtime.config.seed),
  tick: 0,
  playerX: SHOOTER_WIDTH / 2,
  health: runtime.config.player_health,
  shield: runtime.resolved.startingShield,
  invulnerableTicks: 0,
  rescueCharge: clamp(runtime.resolved.rescueCharge, 0, 100),
  rescueHeld: false,
  rescuesUsed: 0,
  grazeCount: 0,
  combo: 0,
  comboClock: 0,
  kills: 0,
  score: 0,
  attackClock: 0,
  attackSequence: 0,
  alignmentTicks: 0,
  companionClocks: runtime.config.companions.map((companion) =>
    companion.trigger === "segment_start" ? Math.max(1, companion.cooldown_ticks) : 0,
  ),
  nextEnemyID: 0,
  nextProjectileID: 0,
  nextPickupID: 0,
  nextEffectID: 0,
  spawnedBoss: false,
  lastRescueTick: 0,
  bossPhaseTick: 0,
  dailyVariant: runtime.dailyVariant,
  enemies: [],
  enemyProjectiles: [],
  playerProjectiles: [],
  pickups: [],
  pickupsCollected: 0,
  lastPickupTick: 0,
  pickupPower: null,
  pickupPowerTicks: 0,
  pressureQuietTicks: 0,
  effects: [],
});

const updateEnemies = (state: ShooterMutableState): void => {
  for (const enemy of state.enemies) {
    if (enemy.health <= 0) continue;
    enemy.age += 1;
    if (enemy.boss) {
      updateBoss(state, enemy);
      continue;
    }
    const spec = state.config.enemies[enemy.specIndex]!;
    moveEnemy(state, enemy, spec);
    enemy.fireClock += 1;
    const interval = encoreInterval(
      spec.fire_interval,
      state.config.encore_level,
      12,
    );
    if (enemy.fireClock >= interval) {
      enemy.fireClock = 0;
      fireEnemy(state, enemy, spec, spec.shot_pattern);
      if (state.config.encore_level >= 2 && (enemy.volley & 1) !== 0) {
        fireEnemySecondary(state, enemy, spec);
      }
      enemy.volley += 1;
    }
    if (enemy.y >= 5_200 - ENEMY_RADIUS && Math.abs(enemy.x - state.playerX) < PLAYER_RADIUS + ENEMY_RADIUS) {
      damagePlayer(state, Math.max(1, spec.contact_damage));
      if (hasTrait(spec, "steal_pickup")) state.rescueCharge = Math.max(0, state.rescueCharge - 20);
      enemy.health = 0;
    }
  }
  removeDefeatedEnemies(state);
};

const snapshot = (state: ShooterMutableState): ShooterSnapshot => ({
  tick: state.tick,
  player_x: state.playerX,
  health: state.health,
  max_health: 3,
  shield: state.shield,
  invulnerable_ticks: state.invulnerableTicks,
  rescue_charge: state.rescueCharge,
  rescues_used: state.rescuesUsed,
  graze_count: state.grazeCount,
  combo: state.combo,
  score: state.score,
  ...(state.pickupPower && state.pickupPowerTicks > 0
    ? {
        pickup_power: state.pickupPower,
        pickup_power_ticks: state.pickupPowerTicks,
      }
    : {}),
  ...(state.dailyVariant ? { daily_variant: state.dailyVariant } : {}),
  enemies: state.enemies.map((enemy) => {
    const spec = enemy.boss ? null : state.config.enemies[enemy.specIndex]!;
    let intent: "fire" | "charge" | "" = "";
    if (enemy.boss && state.config.boss) {
      const stage = state.config.boss.stages[enemy.phase - 1];
      if (stage?.telegraph_ticks && enemy.fireClock >= Math.max(1, stage.fire_interval - stage.telegraph_ticks)) intent = "fire";
    } else if (spec) {
      intent = enemyIntent(enemy, spec, state.config.encore_level);
    }
    return {
      id: enemy.id,
      spec_id: enemy.boss && state.config.boss ? state.config.boss.id : (spec?.id ?? ""),
      chassis: enemy.boss ? "censor-frame" : (spec?.chassis ?? "spam-bot"),
      position: { x: enemy.x, y: enemy.y },
      health: Math.max(0, enemy.health),
      max_health: enemy.maxHealth,
      boss: enemy.boss,
      ...(enemy.phase ? { stage: enemy.phase } : {}),
      ...(intent ? { intent } : {}),
      ...(enemy.marks ? { marks: enemy.marks } : {}),
    };
  }),
  enemy_projectiles: state.enemyProjectiles.map((item) => ({
    id: item.id,
    position: { x: item.x, y: item.y },
    velocity: { x: item.vx, y: item.vy },
    hostile: item.hostile,
    ...(item.kind ? { kind: item.kind } : {}),
    ...(item.radius ? { radius: item.radius } : {}),
    ...(item.width ? { width: item.width } : {}),
    ...(item.health > 0 ? { health: item.health } : {}),
  })),
  player_projectiles: state.playerProjectiles.map((item) => ({
    id: item.id,
    position: { x: item.x, y: item.y },
    velocity: { x: item.vx, y: item.vy },
    hostile: item.hostile,
    ...(item.kind ? { kind: item.kind } : {}),
    ...(item.radius ? { radius: item.radius } : {}),
    ...(item.width ? { width: item.width } : {}),
    ...(item.health > 0 ? { health: item.health } : {}),
  })),
  pickups: state.pickups.map((item) => ({
    id: item.id,
    kind: item.kind,
    position: { x: item.x, y: item.y },
    value: item.value,
  })),
  threats: threatSnapshots(state),
  effects: state.effects.map((item) => ({
    id: item.id,
    kind: item.kind,
    position: { x: item.x, y: item.y },
    ticks: item.ticks,
    ...(item.power ? { power: item.power } : {}),
  })),
});

export const createShooterSimulation = (runtime: ShooterRuntime): ShooterSimulation => {
  const state = createInitialState(runtime);
  let cachedResult: ShooterResult | null = null;

  const step = (input: ShooterInput): ShooterStepEvents => {
    if (state.tick >= state.config.duration_ticks || state.health <= 0) {
      return emptyStepEvents();
    }
    const before = {
      health: state.health,
      shield: state.shield,
      combo: state.combo,
      pickups: state.pickupsCollected,
      rescues: state.rescuesUsed,
      warnings: threatSnapshots(state).length,
      enemyHealth: new Map(state.enemies.map((enemy) => [enemy.id, enemy.health])),
    };
    state.tick += 1;
    if (state.invulnerableTicks > 0) state.invulnerableTicks -= 1;
    let moveLimit = state.config.kit.move_limit;
    if (moveLimit <= 0 || moveLimit > SHOOTER_WIDTH / 2 - PLAYER_RADIUS) moveLimit = SHOOTER_WIDTH / 2 - PLAYER_RADIUS;
    state.playerX = SHOOTER_WIDTH / 2 - moveLimit + input.x * (moveLimit * 2) / 127;
    state.playerX = Math.trunc(state.playerX);
    const rescueActivated = input.rescue && !state.rescueHeld ? activateRescue(state) : false;
    state.rescueHeld = input.rescue;
    spawnWave(state);
    spawnBoss(state);
    updateWeapons(state);
    updateCompanions(state);
    updateEnemies(state);
    updateKitPassives(state);
    updateProjectiles(state);
    updatePickups(state);
    updateEffects(state);
    if (state.pickupPowerTicks > 0) state.pickupPowerTicks -= 1;
    if (state.pickupPowerTicks === 0) state.pickupPower = null;
    if (state.comboClock > 0) state.comboClock -= 1;
    else state.combo = 0;
    if (state.health < 0) state.health = 0;
    const warnings = threatSnapshots(state).length;
    const enemyHitIDs = state.enemies
      .filter(
        (enemy) =>
          (before.enemyHealth.get(enemy.id) ?? enemy.maxHealth) > enemy.health,
      )
      .map((enemy) => enemy.id);
    const enemyDefeatedIDs = state.enemies
      .filter(
        (enemy) =>
          enemy.health <= 0 &&
          (before.enemyHealth.get(enemy.id) ?? enemy.maxHealth) > 0,
      )
      .map((enemy) => enemy.id);
    return {
      pickup: state.pickupsCollected > before.pickups,
      enemyHitIDs,
      enemyDefeatedIDs,
      hit: state.health < before.health,
      shield: state.shield < before.shield && state.health === before.health,
      combo: state.combo > before.combo,
      rescue: rescueActivated || state.rescuesUsed > before.rescues,
      bossWarning: warnings > before.warnings,
    };
  };

  const result = (): ShooterResult | null => {
    const aliveBoss = state.enemies.some((enemy) => enemy.boss && enemy.health > 0);
    const bossDefeated = Boolean(
      state.config.boss && state.spawnedBoss && !aliveBoss,
    );
    if (
      state.health > 0 &&
      state.tick < state.config.duration_ticks &&
      !bossDefeated
    ) {
      return null;
    }
    if (cachedResult) return cachedResult;
    const won = state.health > 0 && (!state.config.boss || !aliveBoss);
    if (won) state.score += state.health * 10 + state.rescueCharge * 2;
    cachedResult = {
      won,
      health: state.health,
      ticks: state.tick,
      kills: state.kills,
      rescues_used: state.rescuesUsed,
      grazes: state.grazeCount,
      score: state.score,
      ...(state.dailyVariant ? { daily_variant: state.dailyVariant } : {}),
      final: snapshot(state),
    };
    return cachedResult;
  };

  return { snapshot: () => snapshot(state), step, result };
};

export const createShooterSimulationFromConfig = (config: ShooterRuntimeConfig): ShooterSimulation =>
  createShooterSimulation(createShooterRuntime(config));
